source/daplink/circ_buf.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file    circular_buffer.c
+ * @brief   Implementation of a circular buffer
+ *
+ * DAPLink Interface Firmware
+ * Copyright (c) 2016-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "circ_buf.h"
+
+#include "cortex_m.h"
+#include "macro.h"
+#include "util.h"
+
+void circ_buf_init(circ_buf_t *circ_buf, uint8_t *buffer, uint32_t size)
+{
+    cortex_int_state_t state;
+    state = cortex_int_get_and_disable();
+
+    circ_buf->buf = buffer;
+    circ_buf->size = size;
+    circ_buf->head = 0;
+    circ_buf->tail = 0;
+
+    cortex_int_restore(state);
+}
+
+void circ_buf_push(circ_buf_t *circ_buf, uint8_t data)
+{
+    cortex_int_state_t state;
+    state = cortex_int_get_and_disable();
+
+    circ_buf->buf[circ_buf->tail] = data;
```

Comment:
In this implementation calling circ_buf_push when the buffer is full is considered a software error. If this happens then it will trigger an assert which will show up on DAPLink's filesystem so this will not go unnoticed.